src/main/java/org/folio/dataimport/handlers/actions/CreateInvoiceEventHandler.java

```diff
@@ -352,26 +360,34 @@ private void linkInvoiceLinesToPoLines(List<InvoiceLine> invoiceLines, Map<Integ
 
   private Future<List<Pair<InvoiceLine, String>>> saveInvoiceLines(List<InvoiceLine> invoiceLines, Map<String, String> okapiHeaders) {
     List<Future<Pair<InvoiceLine, String>>> futures = new ArrayList<>();
-    Semaphore semaphore = new Semaphore(5, Vertx.currentContext().owner());
+
+    if (CollectionUtils.isEmpty(invoiceLines)) {
+      return Future.succeededFuture(new ArrayList<>());
+    }
 
     InvoiceLineHelper helper = new InvoiceLineHelper(okapiHeaders, Vertx.currentContext());
     return Vertx.currentContext()
       .<List<Future<Pair<InvoiceLine, String>>>>executeBlocking(promise -> {
+        Semaphore semaphore = new Semaphore(SEMAPHORE_MAX_ACTIVE_THREADS, Vertx.currentContext().owner());
         for (InvoiceLine invoiceLine : invoiceLines) {
-          var future = helper.createInvoiceLine(invoiceLine)
-            .compose(createdInvoiceLine -> {
-              Pair<InvoiceLine, String> invoiceLineToMsg = Pair.of(createdInvoiceLine, null);
-              return Future.succeededFuture(invoiceLineToMsg);
-            }, err -> {
-              logger.error("Failed to create invoice line {}, {}", invoiceLine, err);
-              Pair<InvoiceLine, String> invoiceLineToMsg = Pair.of(invoiceLine, err.getMessage());
-              return Future.succeededFuture(invoiceLineToMsg);
-            });
-          futures.add(future);
-          semaphore.acquire(() ->
-            future.onComplete(asyncResult -> semaphore.release()));
+          semaphore.acquire(() -> {
+            var future = helper.createInvoiceLine(invoiceLine)
+              .compose(createdInvoiceLine -> {
+                Pair<InvoiceLine, String> invoiceLineToMsg = Pair.of(createdInvoiceLine, null);
+                return Future.succeededFuture(invoiceLineToMsg);
+              }, err -> {
+                logger.error("Failed to create invoice line {}, {}", invoiceLine, err);
+                Pair<InvoiceLine, String> invoiceLineToMsg = Pair.of(invoiceLine, err.getMessage());
+                return Future.succeededFuture(invoiceLineToMsg);
+              })
+              .onComplete(asyncResult -> semaphore.release());
+            futures.add(future);
+            // complete executeBlocking promise when all operations processed
```

Comment:
I think we should use this comment as in other places, because not operations can be processed when we closing promise
// complete executeBlocking promise when all operations started